flocker/node/agents/cinder.py

```diff
@@ -472,43 +472,103 @@ def destroy_volume(self, blockdevice_id):
                 break
             time.sleep(1.0)
 
-    def get_device_path(self, blockdevice_id):
-        # libvirt does not return the correct device path when additional
-        # disks have been attached using a client other than cinder. This is
-        # expected behaviour within Cinder and libvirt
-        # See https://bugs.launchpad.net/cinder/+bug/1387945 and
-        # http://libvirt.org/formatdomain.html#elementsDisks (target section)
-        # However, the correct device is named as a udev symlink which includes
-        # the first 20 characters of the blockedevice_id.
-        device_path = FilePath(
-            "/dev/disk/by-id/virtio-{}".format(blockdevice_id[:20]))
-        if not device_path.exists():
-            # If the device path does not exist, either virtio driver is
-            # not being used (e.g. Rackspace), or the user has modified
-            # their udev rules.  The following code relies on Cinder
-            # returning the correct device path, which appears to work
-            # for Rackspace and will work with virtio if no disks have
-            # been attached outside Cinder.
-            try:
-                cinder_volume = self.cinder_volume_manager.get(blockdevice_id)
-            except CinderNotFound:
-                raise UnknownVolume(blockdevice_id)
+    def _get_device_path_virtio_blk(self, volume):
+        """
+        The virtio_blk driver allows a serial number to be assigned to virtual
+        blockdevices.
+        OpenStack will set a serial number containing the first 20
+        characters of the Cinder block device ID.
+
+        This was introduced in
+        * https://github.com/openstack/nova/commit/3a47c02c58cefed0e230190b4bcef14527c82709  # noqa
+        * https://bugs.launchpad.net/nova/+bug/1004328
+
+        The udev daemon will read the serial number and create a
+        symlink to the canonical virtio_blk device path.
+
+        We do this because libvirt does not return the correct device path when
+        additional disks have been attached using a client other than
+        cinder. This is expected behaviour within Cinder and libvirt See
+        https://bugs.launchpad.net/cinder/+bug/1387945 and
+        http://libvirt.org/formatdomain.html#elementsDisks (target section)
+
+        :param volume: The Cinder ``Volume`` which is attached.
+        :returns: ``FilePath`` of the device created by the virtio_blk
+            driver.
+        """
+        expected_path = FilePath(
+            "/dev/disk/by-id/virtio-{}".format(volume.id[:20])
+        )
+        if expected_path.exists():
+            return expected_path.realpath()
```

Comment:
Is there a reason you chose to return the real device path, rather than the symbolic link?  It would be good to document any reason, because the symbolic link has the advantage of being less ambiguous in the face of concurrent detaching/attaching activity.